src/coreclr/jit/emitloongarch64.cpp

```diff
@@ -0,0 +1,6431 @@
+
+        reg2 = REG_RA;
+    }
+
+    instrDesc* id = emitNewInstr(attr);
+
+    id->idReg1(reg1);
+
+    id->idReg2(reg2);
+
+    id->idIns(ins);
+
+    code_t code = emitInsCode(ins);
+    D_INST_2RI12(code, (reg1 & 0x1f), reg2, imm);
+
+    id->idAddr()->iiaSetInstrEncode(code);
+    id->idAddr()->iiaLclVar.initLclVarAddr(varx, offs);
+    id->idSetIsLclVar();
+
+    id->idCodeSize(4);
+    // dispIns(id);
+    appendToCurIG(id);
+}
+
+/*
+ *  Special notes for `offs`, please see the comment for `emitter::emitIns_S_R`.
+ */
+void emitter::emitIns_R_S(instruction ins, emitAttr attr, regNumber reg1, int varx, int offs)
+{
+    ssize_t imm;
+
+    emitAttr size = EA_SIZE(attr);
+
+#ifdef DEBUG
+    switch (ins)
+    {
+        case INS_ld_b:
+        case INS_ld_bu:
+
+        case INS_ld_h:
+        case INS_ld_hu:
+
+        case INS_ld_w:
+        case INS_ld_wu:
+        case INS_fld_s:
+
+        case INS_ld_d:
+        case INS_fld_d:
+
+            break;
+
+        case INS_lea:
+            assert(size == EA_8BYTE);
+            break;
+
+        default:
+            NYI("emitIns_R_S");
+            return;
+
+    } // end switch (ins)
+#endif
+
+    /* Figure out the variable's frame position */
+    int  base;
+    bool FPbased;
+
+    base = emitComp->lvaFrameAddress(varx, &FPbased);
+    imm  = offs < 0 ? -offs - 8 : base + offs;
+
+    regNumber reg2 = FPbased ? REG_FPBASE : REG_SPBASE;
+    reg2           = offs < 0 ? REG_R21 : reg2;
+    offs           = offs < 0 ? -offs - 8 : offs;
+
+    reg1 = (regNumber)((char)reg1 & 0x1f);
+    code_t code;
+    if ((-2048 <= imm) && (imm < 2048))
+    {
+        if (ins == INS_lea)
+        {
+            ins = INS_addi_d;
+        }
+        code = emitInsCode(ins);
+        D_INST_2RI12(code, reg1, reg2, imm);
+    }
+    else
+    {
+        if (ins == INS_lea)
+        {
+            assert(isValidSimm20(imm >> 12));
+            emitIns_R_I(INS_lu12i_w, EA_PTRSIZE, REG_RA, imm >> 12);
+            ssize_t imm2 = imm & 0xfff;
+            emitIns_R_R_I(INS_ori, EA_PTRSIZE, REG_RA, REG_RA, imm2);
+
+            ins  = INS_add_d;
+            code = emitInsCode(ins);
+            D_INST_add_d(code, reg1, reg2, REG_RA);
+        }
+        else
+        {
+            ssize_t imm3 = imm & 0x800;
+            ssize_t imm2 = imm + imm3;
+            assert(isValidSimm20(imm2 >> 12));
+            emitIns_R_I(INS_lu12i_w, EA_PTRSIZE, REG_RA, imm2 >> 12);
+
+            emitIns_R_R_R(INS_add_d, attr, REG_RA, REG_RA, reg2);
+
+            imm2 = imm2 & 0x7ff;
+            code = emitInsCode(ins);
+            D_INST_2RI12(code, reg1 /* & 0x1f*/, REG_RA, imm3 ? imm2 - imm3 : imm2);
+        }
+    }
+
+    instrDesc* id = emitNewInstr(attr);
+
+    id->idReg1(reg1);
+
+    id->idIns(ins);
+
+    id->idAddr()->iiaSetInstrEncode(code);
+    id->idAddr()->iiaLclVar.initLclVarAddr(varx, offs);
+    id->idSetIsLclVar();
+
+    id->idCodeSize(4);
+    // dispIns(id);
```

Comment:
I guess I asked this elsewhere, but why is dispIns not used? Where is the new instruction displayed?

Reply:
Can you show the output of `COMPlus_JitDisasm` for a simple hello-world function?

Reply:
@BruceForstall 

> I guess I asked this elsewhere, but why is dispIns not used? Where is the new instruction displayed?
Same comment with https://github.com/dotnet/runtime/pull/62843#discussion_r828673607

Should I delete this comment `// dispIns(id);`  ?

> Can you show the output of `COMPlus_JitDisasm` for a simple hello-world function?
Here is a similar output
```
{ Start Jitting Method    0 System.Runtime.CompilerServices.CastHelpers:StelemRef(System.Array,int,System.Object) (MethodHash=1b5e9e9c) MinOpts; Assembly listing for method System.Runtime.CompilerServices.CastHelpers:StelemRef(System.Array,int,System.Object)
; Emitting BLENDED_CODE for generic LOONGARCH64 CPU - Unix
; debuggable code
; fp based frame
; fully interruptible
; No PGO data 
; Final local variable assignments
;
;  V00 arg0         [V00    ] (  1,  1   )     ref  ->  [fp+88H]   do-not-enreg[] class-hnd
;  V01 arg1         [V01    ] (  1,  1   )     int  ->  [fp+84H]   do-not-enreg[]
;  V02 arg2         [V02    ] (  1,  1   )     ref  ->  [fp+78H]   do-not-enreg[] class-hnd
;  V03 loc0         [V03    ] (  1,  1   )   byref  ->  [fp+70H]   do-not-enreg[] must-init
;  V04 loc1         [V04    ] (  1,  1   )    long  ->  [fp+68H]   do-not-enreg[]
;  V05 loc2         [V05    ] (  1,  1   )    bool  ->  [fp+64H]   do-not-enreg[]
;  V06 loc3         [V06    ] (  1,  1   )    bool  ->  [fp+60H]   do-not-enreg[]
;  V07 loc4         [V07    ] (  1,  1   )    bool  ->  [fp+5CH]   do-not-enreg[]
;# V08 OutArgs      [V08    ] (  1,  1   )  lclBlk ( 0) [sp+00H]   do-not-enreg[] "OutgoingArgSpace"
;  V09 tmp1         [V09    ] (  1,  1   )     ref  ->  [fp+50H]   do-not-enreg[] must-init class-hnd "impSpillStackEnsure"
;  V10 tmp2         [V10    ] (  1,  1   )    long  ->  [fp+48H]   do-not-enreg[] "impSpillStackEnsure"
;  V11 tmp3         [V11    ] (  1,  1   )    long  ->  [fp+40H]   do-not-enreg[] "impSpillStackEnsure"
;  V12 tmp4         [V12    ] (  1,  1   )    long  ->  [fp+38H]   do-not-enreg[] "impSpillStackEnsure"
;  V13 tmp5         [V13    ] (  1,  1   )     ref  ->  [fp+30H]   do-not-enreg[] must-init class-hnd "impSpillStackEnsure"
;  V14 tmp6         [V14    ] (  1,  1   )  struct ( 8) [fp+28H]   do-not-enreg[SF] must-init "struct address for call/obj"
;  V15 tmp7         [V15    ] (  1,  1   )     ref  ->  [fp+20H]   do-not-enreg[] must-init class-hnd "impSpillStackEnsure"
;  V16 tmp8         [V16    ] (  1,  1   )     int  ->  [fp+1CH]   do-not-enreg[] "impSpillStackEnsure"
;
; Lcl frame size = 128
DEBUG: LOONGARCH64, frameType:1


G_M24931_IG01:              ;; offset=0000H
   0xff7d00c410   addi.d  sp, sp, -144
   0xff7d00c414   st.d  fp, sp, 0
   0xff7d00c418   st.d  ra, sp, 8
   0xff7d00c41c   ori  fp, sp, 0x0
  0xff7d00c420   addi.d  a3, fp, 32
   0xff7d00c424   addi.w  a4, zero, 5
   0xff7d00c428   st.d  zero, a3, 8
   0xff7d00c42c   st.d  zero, a3, 0
   0xff7d00c430   addi.d  a4, a4, -1
   0xff7d00c434   addi.d  a3, a3, 16
   0xff7d00c438   bne  a4, zero, 0xff7d00c428
   0xff7d00c43c   st.d  zero, a3, 0
   0xff7d00c440   st.d  a0, fp, 136
   0xff7d00c444   st.w  a1, fp, 132
   0xff7d00c448   st.d  a2, fp, 120
                        ;; bbWeight=1    PerfScore 0.00
G_M24931_IG02:              ;; offset=003CH
   0xff7d00c44c   lu12i.w  a0, 0x7ce40
   0xff7d00c450   ori  a0, a0, 0x588
   0xff7d00c454   lu32i.d  a0, 0xff
   0xff7d00c458   ld.w  a0, a0, 0
   0xff7d00c45c   beqz  a0, 0xff7d00c470
                        ;; bbWeight=1    PerfScore 0.00
G_M24931_IG03:              ;; offset=0050H
   0xff7d00c460   lu12i.w  a0, 0xf70db
   0xff7d00c464   ori  a0, a0, 0x40
   0xff7d00c468   lu32i.d  a0, 0xff
   0xff7d00c46c   jirl  ra, a0, 0  #CORINFO_HELP_DBG_IS_JUST_MY_CODE
                        ;; bbWeight=0.50 PerfScore 0.00
G_M24931_IG04:              ;; offset=0060H
   0xff7d00c470   andi  zero, zero, 0x0
   0xff7d00c474   ld.d  a1, fp, 136
   0xff7d00c478   lu12i.w  a0, 0x7d124
   0xff7d00c47c   ori  a0, a0, 0x928
   0xff7d00c480   lu32i.d  a0, 0xff
   0xff7d00c484   lu12i.w  a2, 0x7d004
   0xff7d00c488   ori  a2, a2, 0x18
   0xff7d00c48c   lu32i.d  a2, 0xff
   0xff7d00c490   jirl  ra, a2, 0  #Internal.Runtime.CompilerServices.Unsafe:As(System.Object):System.__Canon
```

This is more  convenient for our debuging.

Thanks.